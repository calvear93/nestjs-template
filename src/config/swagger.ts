import { INestApplication } from '@nestjs/common';
import { SwaggerModule, DocumentBuilder } from '@nestjs/swagger';

export default (app : INestApplication) =>
{
    const config = new DocumentBuilder()
        .setTitle(process.env.TITLE)
        .setDescription(process.env.DESCRIPTION)
        .setVersion(process.env.VERSION)
        .addBearerAuth()
        .build();

    const document = SwaggerModule.createDocument(app, config);

    SwaggerModule.setup('', app, document);
};
