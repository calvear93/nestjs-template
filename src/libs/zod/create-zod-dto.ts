import { type ZodTypeAny, type ZodType, type z } from 'zod';
import {
	type OpenAPIObject,
	type ReferenceObject,
	type SchemaObject,
} from '@nestjs/swagger/dist/interfaces/open-api-spec.interface';
import { zodToJsonSchema } from './json-schema.factory.js';

export type Constructor<T, Arguments extends unknown[] = any[]> = new (
	...arguments_: Arguments
) => T;

export interface ZodDto<T = any, S extends ZodType = ZodTypeAny>
	extends Constructor<T> {
	readonly schema: S;
	get jsonSchema(): SchemaObject;
}

const registered: [
	name: string,
	genSchema: () => SchemaObject | ReferenceObject,
][] = [];

/**
 * Creates a DTO from Zod schema,
 * with schema and jsonSchema
 * static properties.
 *
 * @param schema - zod schema
 * @param name - json schema name for register
 *
 * @example
 * ```ts
 *	// sample.dto.ts
 *	import { z } from 'zod';
 *	import { createZodDto } from '@libs/zod';
 *
 *	const SampleSchema = z.object({ id: z.number(), name: z.string() });
 *
 *	export class SampleDto extends createZodDto(SampleSchema, 'Sample') {}
 *
 *	// sample.controller.ts
 *	import { SampleDto } from './sample.dto.ts';
 *
 *	\@Controller('sample')
 *	export class SampleController {
 *		\@Post()
 *		\@ApiBody({ schema: SampleDto.jsonSchema })
 *		sample(\@Body() demo: SampleDto): any { ... }
 *	}
 *
 *	// sample.parser.ts
 *	import { SampleDto } from './sample.dto.ts';
 *
 *	export const parseSampleDto = (input: unknown) => {
 *		return SampleDto.schema.parse(input);
 *	}
 * ```
 */
export const createZodDto = <T extends ZodType>(
	schema: T,
	name?: string,
): ZodDto<z.infer<T>, T> => {
	let jsonSchema: SchemaObject | undefined;
	const getSchema = () =>
		jsonSchema ?? (jsonSchema = zodToJsonSchema(schema));

	if (name) {
		registered.push([name, getSchema]);
	}

	return class {
		static readonly schema = schema;

		static get jsonSchema() {
			return getSchema();
		}
	};
};

/**
 * Register Zod DTOs schemas to
 * OpenApi Swagger document.
 *
 * @example
 * ```ts
 *	import { ... } from '...';
 *	import { registerDtoSchemas } from '@libs/zod';
 *
 *	const app = await NestFactory.create(AppModule);
 *
 *	const config = new DocumentBuilder().build();
 *	const document = SwaggerModule.createDocument(app, config);
 *
 *	registerDtoSchemas(document);
 * ```
 */
export const registerDtoSchemas = (openApi: OpenAPIObject): OpenAPIObject => {
	openApi.components ??= {};
	openApi.components.schemas ??= {};

	for (const [name, genSchema] of registered) {
		openApi.components.schemas[name] = genSchema();
	}

	return openApi;
};
