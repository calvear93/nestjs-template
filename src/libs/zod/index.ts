export { createZodDto, registerDtoSchemas } from './create-zod-dto.js';
export { ZodValidationPipe } from './zod.pipe.js';
