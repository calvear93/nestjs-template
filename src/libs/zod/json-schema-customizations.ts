import type {
	OpenAPIObject,
	SchemaObject,
} from '@nestjs/swagger/dist/interfaces/open-api-spec.interface';
import z from 'zod';

type JsonSchemaCustomization =
	| z.core.JSONSchemaMeta
	| (<Z extends z.core.$ZodTypes>(
			jsonSchema: z.core.JSONSchema.BaseSchema,
			zodSchema: Z,
	  ) => z.core.JSONSchemaMeta);
type JsonSchemaCustomizations = Record<string, JsonSchemaCustomization>;

// global JSON schema registry for OpenAPI
const registered: [name: string, jsonSchema: SchemaObject][] = [];

export const JsonSchemaCustomizations: {
	readonly formats: JsonSchemaCustomizations;
	readonly types: JsonSchemaCustomizations;
} = {
	formats: {
		regex: (_, { _zod: { def } }: any) => ({
			pattern: def.pattern?.toString(),
			type: 'string',
		}),
	},
	types: {
		void: {
			default: null,
			nullable: true,
			readOnly: true,
			type: 'null',
		},
		any: {
			nullable: true,
		},
		bigint: {
			format: 'int64',
			type: 'integer',
		},
		date: {
			format: 'date-time',
			type: 'string',
		},
		nan: {
			nullable: true,
			type: 'number',
		},
		never: {
			not: {},
			readOnly: true,
		},
		nullable: {
			nullable: true,
		},
		regex: {
			pattern: 'date-time',
			type: 'string',
		},
		symbol: {
			default: null,
			type: 'string',
		},
		undefined: {
			readOnly: true,
		},
		map: (_, { _zod: { def } }: any) => ({
			additionalProperties: def.valueType.def,
			type: 'object',
		}),
		set: (_, { _zod: { def } }: any) => ({
			items: def.valueType.def,
			type: 'object',
		}),
		tuple: ({ prefixItems }) => ({
			maxItems: prefixItems?.length,
			minItems: prefixItems?.length,
		}),
	},
};

type ToJSONSchemaParams = Parameters<typeof z.toJSONSchema>[1];

const getOrCall = (
	customization: JsonSchemaCustomization,
	jsonSchema: z.core.JSONSchema.BaseSchema,
	zodSchema: z.core.$ZodTypes,
) => {
	if (typeof customization === 'function') {
		return customization(jsonSchema, zodSchema);
	}
	return customization;
};

export const applyJsonSchemaCustomizations = (): ToJSONSchemaParams => {
	return {
		override: (context) => {
			const { format, type } = context.zodSchema._zod.def as any;

			// applies type customizations
			const customType = getOrCall(
				JsonSchemaCustomizations.types[type],
				context.jsonSchema,
				context.zodSchema,
			);
			if (customType) Object.assign(context.jsonSchema, customType);

			// applies format customizations
			const customFormat = getOrCall(
				JsonSchemaCustomizations.formats[format],
				context.jsonSchema,
				context.zodSchema,
			);
			if (customFormat) Object.assign(context.jsonSchema, customFormat);
		},
		unrepresentable: 'any',
	};
};

export const toJSONSchema = (schema: z.ZodTypeAny, schemaName?: string) => {
	const jsonSchema = z.toJSONSchema(
		schema,
		applyJsonSchemaCustomizations(),
	) as SchemaObject;

	// registers global JSON schema for OpenAPI only if schemaName is provided
	if (schemaName) registered.push([schemaName, jsonSchema]);

	return jsonSchema;
};

/**
 * Register Zod DTOs schemas to OpenApi Swagger document.
 *
 * @example
 * ```ts
 *	import { ... } from '...';
 *	import { registerDtoSchemas } from '#libs/zod';
 *
 *	const app = await NestFactory.create(AppModule);
 *
 *	const config = new DocumentBuilder().build();
 *	const document = SwaggerModule.createDocument(app, config);
 *
 *	registerDtoSchemas(document);
 * ```
 */
export const registerDtoOpenApiSchemas = (
	openApi: OpenAPIObject,
): OpenAPIObject => {
	openApi.components ??= {};
	openApi.components.schemas ??= {};

	for (const [name, schema] of registered) {
		openApi.components.schemas[name] = schema;
	}

	return openApi;
};
