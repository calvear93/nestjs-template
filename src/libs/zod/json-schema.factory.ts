import {
	z,
	type AnyZodObject,
	type ZodOptionalType,
	type ZodTypeAny,
} from 'zod';
import { type SchemaObject } from '@nestjs/swagger/dist/interfaces/open-api-spec.interface';

interface ParsingArgs<T> {
	zodRef: T;
	schema: SchemaObject;
}

const isOptional = (item: ZodTypeAny): item is ZodOptionalType<any> => {
	return (
		item.isOptional() ||
		item instanceof z.ZodDefault ||
		item instanceof z.ZodNever ||
		item._def.typeName === 'ZodDefault'
	);
};

const iterateZodObject = ({ zodRef }: ParsingArgs<AnyZodObject>) => {
	const keys = Object.keys(zodRef.shape) as string[];

	return Object.fromEntries(
		keys.map((key: string) => [key, zodToJsonSchema(zodRef.shape[key])]),
	);
};

const toObject = ({
	zodRef,
	schema,
}: ParsingArgs<z.ZodObject<any>>): SchemaObject => {
	schema.type = 'object';
	schema.properties = iterateZodObject({ zodRef, schema });

	const required = Object.keys(zodRef.shape).filter((key) => {
		const item = zodRef.shape[key];
		return !isOptional(item);
	}) as string[];

	if (required.length > 0) schema.required = required;

	return schema;
};

const toRecord = ({
	zodRef,
	schema,
}: ParsingArgs<z.ZodRecord>): SchemaObject => {
	schema.type = 'object';
	schema.additionalProperties =
		zodRef._def.valueType instanceof z.ZodUnknown
			? {}
			: zodToJsonSchema(zodRef._def.valueType);

	return schema;
};

const toBoolean = ({ schema }: ParsingArgs<z.ZodBoolean>): SchemaObject => {
	schema.type = 'boolean';

	return schema;
};

const toString = ({
	zodRef,
	schema,
}: ParsingArgs<z.ZodString>): SchemaObject => {
	schema.type = 'string';

	const { checks = [] } = zodRef._def;

	for (const item of checks) {
		switch (item.kind) {
			case 'email':
				schema.format = 'email';
				break;
			case 'uuid':
				schema.format = 'uuid';
				break;
			case 'cuid':
				schema.format = 'cuid';
				break;
			case 'url':
				schema.format = 'uri';
				break;
			case 'length':
				schema.minLength = item.value;
				schema.maxLength = item.value;
				break;
			case 'max':
				schema.maxLength = item.value;
				break;
			case 'min':
				schema.minLength = item.value;
				break;
			case 'regex':
				schema.pattern = item.regex.source;
				break;
			default:
				break;
		}
	}

	return schema;
};

const toLiteral = ({
	schema,
	zodRef,
}: ParsingArgs<z.ZodLiteral<ZodTypeAny>>): SchemaObject => {
	schema.type = typeof zodRef._def.value;
	schema.enum = [zodRef._def.value];

	return schema;
};

const toNumber = ({
	zodRef,
	schema,
}: ParsingArgs<z.ZodNumber>): SchemaObject => {
	schema.type = 'number';

	const { checks = [] } = zodRef._def;

	for (const item of checks) {
		switch (item.kind) {
			case 'max':
				schema.maximum = item.value;
				if (!item.inclusive) schema.exclusiveMaximum = true;
				break;
			case 'min':
				schema.minimum = item.value;
				if (!item.inclusive) schema.exclusiveMinimum = true;
				break;
			case 'int':
				schema.type = 'integer';
				break;
			case 'multipleOf':
				schema.multipleOf = item.value;
				break;
			default:
				break;
		}
	}

	return schema;
};

const toBigInt = ({ schema }: ParsingArgs<z.ZodBigInt>): SchemaObject => {
	schema.type = 'integer';
	schema.format = 'int64';

	return schema;
};

const toNaN = ({ schema }: ParsingArgs<z.ZodNumber>): SchemaObject => {
	schema.type = 'number';
	schema.nullable = true;
	schema.readOnly = true;
	schema.default = Number.NaN;

	return schema;
};

const toDate = ({ schema }: ParsingArgs<z.ZodDate>): SchemaObject => {
	schema.type = 'string';
	schema.format = 'date-time';

	return schema;
};

const toNull = ({ schema }: ParsingArgs<z.ZodNull>): SchemaObject => {
	schema.type = 'number';
	schema.format = 'null';
	schema.nullable = true;
	schema.default = null;

	return schema;
};

const toVoidOrNever = ({ schema }: ParsingArgs<z.ZodNever>): SchemaObject => {
	schema.readOnly = true;

	return schema;
};

const toArray = ({
	schema,
	zodRef,
}: ParsingArgs<z.ZodArray<ZodTypeAny>>): SchemaObject => {
	schema.type = 'array';
	schema.items = zodToJsonSchema(zodRef.element);

	if (zodRef._def.exactLength !== null) {
		schema.minItems = zodRef._def.exactLength.value;
		schema.maxItems = zodRef._def.exactLength.value;
	}

	if (zodRef._def.minLength !== null)
		schema.minItems = zodRef._def.minLength.value;
	if (zodRef._def.maxLength !== null)
		schema.maxItems = zodRef._def.maxLength.value;

	return schema;
};

const toTuple = <
	T extends [ZodTypeAny, ...ZodTypeAny[]],
	Rest extends ZodTypeAny | null,
>({
	schema,
	zodRef,
}: ParsingArgs<z.ZodTuple<T, Rest>>): SchemaObject => {
	schema.type = 'array';
	schema.items = [] as any;
	(schema as any).prefixItems = zodRef._def.items.map((t) =>
		zodToJsonSchema(t),
	);

	return schema;
};

const toEnum = ({
	schema,
	zodRef,
}: ParsingArgs<z.ZodEnum<never> | z.ZodNativeEnum<never>>): SchemaObject => {
	schema.type = typeof Object.values(zodRef._def.values)[0];
	schema.enum = Object.values(zodRef._def.values);

	return schema;
};

const toIntersection = ({
	schema,
	zodRef,
}: ParsingArgs<
	z.ZodIntersection<z.ZodTypeAny, z.ZodTypeAny>
>): SchemaObject => {
	schema.allOf = [
		zodToJsonSchema(zodRef._def.left),
		zodToJsonSchema(zodRef._def.right),
	];

	return schema;
};

const toUnion = ({
	schema,
	zodRef,
}: ParsingArgs<
	z.ZodUnion<[z.ZodTypeAny, ...z.ZodTypeAny[]]>
>): SchemaObject => {
	schema.oneOf = (
		zodRef as z.ZodUnion<[z.ZodTypeAny, ...z.ZodTypeAny[]]>
	)._def.options.map((schema) => zodToJsonSchema(schema));

	return schema;
};

const toDiscriminatedUnion = ({
	schema,
	zodRef,
}: ParsingArgs<
	z.ZodDiscriminatedUnion<string, z.ZodDiscriminatedUnionOption<string>[]>
>): SchemaObject => {
	schema.discriminator = {
		propertyName: (
			zodRef as z.ZodDiscriminatedUnion<
				string,
				z.ZodDiscriminatedUnionOption<string>[]
			>
		)._def.discriminator,
	};

	schema.oneOf = [
		...(
			zodRef as z.ZodDiscriminatedUnion<
				string,
				z.ZodDiscriminatedUnionOption<string>[]
			>
		)._def.options.values(),
	].map((schema) => zodToJsonSchema(schema));

	return schema;
};

const toOptionalNullable = ({
	schema,
	zodRef,
}: ParsingArgs<
	z.ZodOptional<ZodTypeAny> | z.ZodNullable<ZodTypeAny>
>): SchemaObject => {
	return { ...schema, ...zodToJsonSchema(zodRef.unwrap()) };
};

const toDefault = ({
	schema,
	zodRef,
}: ParsingArgs<z.ZodDefault<ZodTypeAny>>): SchemaObject => {
	return {
		...schema,
		default: zodRef._def.defaultValue(),
		...zodToJsonSchema(zodRef._def.innerType),
	};
};

const toTransformer = ({
	zodRef,
	schema,
}: ParsingArgs<
	z.ZodTransformer<never> | z.ZodEffects<never>
>): SchemaObject => {
	return { ...schema, ...zodToJsonSchema(zodRef._def.schema) };
};

const catchAll = ({ schema }: ParsingArgs<ZodTypeAny>): SchemaObject => {
	return schema;
};

const PARSERS_LOOKUP = {
	ZodObject: toObject,
	ZodRecord: toRecord,
	ZodBoolean: toBoolean,
	ZodString: toString,
	ZodLiteral: toLiteral,
	ZodNumber: toNumber,
	ZodBigInt: toBigInt,
	ZodNaN: toNaN,
	ZodDate: toDate,
	ZodNull: toNull,
	ZodUndefined: toVoidOrNever,
	ZodVoid: toVoidOrNever,
	ZodNever: toVoidOrNever,
	ZodArray: toArray,
	ZodTuple: toTuple,
	ZodEnum: toEnum,
	ZodNativeEnum: toEnum,
	ZodIntersection: toIntersection,
	ZodUnion: toUnion,
	ZodDiscriminatedUnion: toDiscriminatedUnion,
	ZodOptional: toOptionalNullable,
	ZodNullable: toOptionalNullable,
	ZodDefault: toDefault,
	ZodTransformer: toTransformer,
	ZodEffects: toTransformer,
	ZodMap: catchAll,
	ZodFunction: catchAll,
	ZodLazy: catchAll,
	ZodPromise: catchAll,
	ZodAny: catchAll,
	ZodUnknown: catchAll,
};

const hasLookupParser = (key: string): key is keyof typeof PARSERS_LOOKUP => {
	return key in PARSERS_LOOKUP;
};

/**
 * Creates a JSON Schema from Zod schema.
 * @see https://github.com/anatine/zod-plugins
 *
 * @param zodRef - zod schema
 */
export function zodToJsonSchema(zodRef: ZodTypeAny): SchemaObject {
	const {
		isNullable,
		description,
		_def: { typeName },
	} = zodRef;

	const schema: SchemaObject = {};

	if (isNullable()) schema.nullable = true;
	if (description) schema.description = description;

	if (hasLookupParser(typeName)) {
		return PARSERS_LOOKUP[typeName]({
			zodRef: zodRef as any,
			schema,
		});
	}

	return schema;
}
