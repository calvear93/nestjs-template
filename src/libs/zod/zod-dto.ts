import {
	type OpenAPIObject,
	type ReferenceObject,
	type SchemaObject,
} from '@nestjs/swagger/dist/interfaces/open-api-spec.interface.ts';
import {
	type AnyZodObject,
	type AnyZodTuple,
	type RawCreateParams,
	type ZodArray,
	type ZodObject,
	type ZodRawShape,
	type ZodTypeAny,
	z,
} from 'zod';
import { zodToJsonSchema } from './json-schema.factory.ts';

const registered: [name: string, jsonSchema: ReferenceObject | SchemaObject][] =
	[];

type InferTuple<Tuple extends [...ZodTypeAny[]]> = {
	[Index in keyof Tuple]: z.infer<Tuple[Index]>;
};
type InferArray<Arr extends [...ZodTypeAny[]]> = z.infer<Arr[0]>[];

/**
 * Creates a DTO from Zod shape,
 * with schema and jsonSchema
 * static properties.
 *
 * @param shape - zod shape
 * @param config - zod type config
 *
 * @example
 * ```ts
 *	// sample.dto.ts
 *	import { z } from 'zod';
 *	import { ZodDto } from #libs/zod';
 *
 *	export class SampleDto extends ZodDto({
 *		id: z.number(),
 *		name: z.string()
 *	}) {}
 *
 *	SampleDto.registerOpenApi();
 *
 *	// sample.controller.ts
 *	import { SampleDto } from './sample.dto.ts';
 *
 *	\@Controller('sample')
 *	export class SampleController {
 *		\@Post()
 *		\@ApiBody({ schema: SampleDto.jsonSchema })
 *		sample(\@Body() demo: SampleDto): any { ... }
 *	}
 *
 *	// sample.parser.ts
 *	import { SampleDto } from './sample.dto.ts';
 *
 *	export const parseSampleDto = (input: unknown) => {
 *		return SampleDto.schema.parse(input);
 *	}
 * ```
 */
export const ZodDto = <
	S extends ZodRawShape,
	Z extends AnyZodObject = ZodObject<S>,
>(
	shape: S,
	config?: RawCreateParams,
): ZodObjectDto<Z> => {
	const schema = z.object(shape, config);

	return class {
		constructor(fields?: Partial<z.infer<Z>>) {
			if (fields) {
				Object.assign(this, schema.parse(fields));
			}
		}

		static registerOpenApi() {
			registered.push([this.name, this.jsonSchema]);
		}

		static readonly jsonSchema = zodToJsonSchema(schema);

		static readonly schema = schema as unknown as Z;
	};
};

/**
 * Creates a DTO from Zod array or tuple,
 * with schema and jsonSchema
 * static properties.
 *
 * @param items - zod types list
 * @param config - zod type config
 *
 * @example
 * ```ts
 *	// sample.dto.ts
 *	import { z } from 'zod';
 *	import { ZodIterableDto } from '@zod';
 *
 *	export class SampleDtoIterable extends ZodIterableDto([
 *		z.number(),
 *		z.string()
 *	]) {}
 *
 *	SampleDtoIterable.registerOpenApi();
 *
 *	// sample.controller.ts
 *	import { SampleDtoIterable } from './sample.dto.ts';
 *
 *	\@Controller('sample')
 *	export class SampleController {
 *		\@Post()
 *		\@ApiBody({ schema: SampleDtoIterable.jsonSchema })
 *		sample(\@Body() demo: SampleDtoIterable): any { ... }
 *	}
 *
 *	// sample.parser.ts
 *	import { SampleDtoIterable } from './sample.dto.ts';
 *
 *	export const parseSampleDtoIterable = (input: unknown) => {
 *		return SampleDtoIterable.schema.parse(input);
 *	}
 * ```
 */
export const ZodIterableDto = <
	T extends [ZodTypeAny, ...ZodTypeAny[]],
	Z extends AnyZodTuple | ZodArray<ZodTypeAny> = T['length'] extends 1
		? ZodArray<z.ZodTypeAny>
		: AnyZodTuple,
	I = T['length'] extends 1 ? InferArray<T> : InferTuple<T>,
>(
	items: T,
	config?: RawCreateParams,
): ZodIterableDto<Z, I> => {
	const schema =
		items.length === 1 ? z.array(items[0], config) : z.tuple(items, config);

	return class extends Array {
		constructor(items?: I) {
			super();
			if (items) {
				this.push(...schema.parse(items));
			}
		}

		static registerOpenApi() {
			registered.push([this.name, this.jsonSchema]);
		}

		static readonly jsonSchema = zodToJsonSchema(schema);

		static readonly schema = schema as unknown as Z;
	};
};

/**
 * Register Zod DTOs schemas to
 * OpenApi Swagger document.
 *
 * @example
 * ```ts
 *	import { ... } from '...';
 *	import { registerDtoSchemas } from '@zod';
 *
 *	const app = await NestFactory.create(AppModule);
 *
 *	const config = new DocumentBuilder().build();
 *	const document = SwaggerModule.createDocument(app, config);
 *
 *	registerDtoSchemas(document);
 * ```
 */
export const registerDtoOpenApiSchemas = (
	openApi: OpenAPIObject,
): OpenAPIObject => {
	openApi.components ??= {};
	openApi.components.schemas ??= {};

	for (const [name, schema] of registered) {
		openApi.components.schemas[name] = schema;
	}

	return openApi;
};

/**
 * Zod DTO base.
 */
export interface ZodDto<Z extends ZodTypeAny = ZodTypeAny> {
	readonly jsonSchema: SchemaObject;
	registerOpenApi: () => void;
	readonly schema: Z;
}

/**
 * Zod DTO from object/shape.
 */
export interface ZodObjectDto<Z extends ZodTypeAny = ZodTypeAny, T = z.infer<Z>>
	extends ZodDto<Z> {
	new (fields?: Partial<T>): T;
}

/**
 * Zod DTO from array/tuple.
 */
export interface ZodIterableDto<
	Z extends ZodTypeAny = ZodTypeAny,
	I = InferArray<any> | InferTuple<any>,
	T = I extends (infer A)[] ? A : I extends [...infer B] ? B : any[],
> extends ZodDto<Z> {
	new (items?: I): T[];
}
