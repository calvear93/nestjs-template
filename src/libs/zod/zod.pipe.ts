import {
	BadRequestException,
	type ArgumentMetadata,
	type PipeTransform,
} from '@nestjs/common';
import { type ZodDto } from './create-zod-dto';

const isZodDto = (dto: any): dto is ZodDto => {
	return !!dto.schema;
};

/**
 * Validates Zod DTO schemas
 * on controller input.
 */
export class ZodValidationPipe implements PipeTransform {
	public transform(value: unknown, { metatype }: ArgumentMetadata): unknown {
		if (isZodDto(metatype)) {
			const result = metatype.schema.safeParse(value);

			if (result.success) return result.data;

			const message = result.error.errors.map(
				({ path, message }) => `${path.join('.')}: ${message}`,
			);

			throw new BadRequestException(message);
		}

		return value;
	}
}
