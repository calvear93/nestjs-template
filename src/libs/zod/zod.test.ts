import { z } from 'zod';
import { describe, expect, test } from 'vitest';
import { type OpenAPIObject } from '@nestjs/swagger';
import { BadRequestException } from '@nestjs/common';
import { ZodValidationPipe } from './zod.pipe.ts';
import { createZodDto, registerDtoSchemas } from './create-zod-dto.ts';

describe(ZodValidationPipe, () => {
	// tests
	test('register DTO schemas', () => {
		const openApi = {} as OpenAPIObject;

		class Dto extends createZodDto({ id: z.number() }) {}
		Dto.register();
		registerDtoSchemas(openApi);

		const { components } = openApi;

		expect(components).toBeDefined();
		expect(components?.schemas?.[Dto.name]).toBeDefined();
	});

	test('DTO generates JSON schema', () => {
		const dto = createZodDto({
			arr: z.array(z.boolean()),
			bigInt: z.bigint(),
			bool: z.boolean().nullable(),
			date: z.date(),
			discriminatedUnion: z.discriminatedUnion('status', [
				z.object({ data: z.string(), status: z.literal('success') }),
				z.object({
					error: z.instanceof(Error),
					status: z.literal('failed'),
				}),
			]),
			email: z.string().email(),
			enum: z.enum(['1', '2', '3']),
			id: z.number().max(10).min(2).default(5),
			intersection: z.number().and(z.string()),
			nan: z.nan(),
			null: z.null(),
			record: z.record(z.string(), z.number()),
			str: z.string().max(10).min(2),
			tuple: z.tuple([z.number(), z.string()]),
			union: z.number().or(z.string()),
		});

		expect(dto.jsonSchema).toBeDefined();
	});

	describe('validation pipe', () => {
		const _pipe = new ZodValidationPipe();
		const _dto = createZodDto({
			number: z.number(),
			string: z.string(),
		});
		const metadata = { metatype: _dto } as any;

		test('parses correct input', () => {
			const input = { number: 1, string: 'str' };

			const result = _pipe.transform(input, metadata);

			expect(result).toStrictEqual(input);
		});

		test('throws BadRequestException on bad input', () => {
			const input = { id: 1 };

			expect(() => _pipe.transform(input, metadata)).toThrow(
				BadRequestException,
			);
		});

		test('when metatype is not ZodDTO input is bypassed', () => {
			const input = { id: 1 };

			const result = _pipe.transform(input, { metatype: {} } as any);

			expect(result).toStrictEqual(input);
		});
	});
});
