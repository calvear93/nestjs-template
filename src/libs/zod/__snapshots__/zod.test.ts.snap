// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ZodValidationPipe > DTO generates JSON schema 1`] = `
{
  "properties": {
    "any": {
      "nullable": true,
    },
    "array": {
      "items": {
        "type": "number",
      },
      "type": "array",
    },
    "arrayFrom5": {
      "items": {
        "type": "number",
      },
      "minItems": 5,
      "type": "array",
    },
    "arrayNonEmpty": {
      "items": {
        "type": "number",
      },
      "minItems": 1,
      "type": "array",
    },
    "arrayOf5": {
      "items": {
        "type": "number",
      },
      "maxItems": 5,
      "minItems": 5,
      "type": "array",
    },
    "boolean": {
      "type": "boolean",
    },
    "coerceDate": {
      "format": "date-time",
      "nullable": true,
      "type": "string",
    },
    "coerceToBool": {
      "nullable": true,
      "type": "boolean",
    },
    "coerceToString": {
      "nullable": true,
      "type": "string",
    },
    "cuid": {
      "example": "cjld2cjxh0000qzrmn831i7rn",
      "format": "cuid",
      "type": "string",
    },
    "cuid2": {
      "example": "tz4a98xxat96iws9zmbrgj3a",
      "format": "cuid2",
      "type": "string",
    },
    "datetime": {
      "format": "date-time",
      "type": "string",
    },
    "default": {
      "default": "hello world",
      "type": "string",
    },
    "discriminated": {
      "discriminator": {
        "propertyName": "status",
      },
      "oneOf": [
        {
          "properties": {
            "data": {
              "type": "string",
            },
            "status": {
              "enum": [
                "success",
              ],
              "type": "string",
            },
          },
          "required": [
            "status",
            "data",
          ],
          "type": "object",
        },
        {
          "properties": {
            "message": {
              "type": "string",
            },
            "status": {
              "enum": [
                "failed",
              ],
              "type": "string",
            },
          },
          "required": [
            "status",
            "message",
          ],
          "type": "object",
        },
      ],
    },
    "email": {
      "example": "sample@email.com",
      "format": "email",
      "type": "string",
    },
    "emoji": {
      "example": "ðŸ˜œ",
      "type": "string",
    },
    "endsWith": {
      "type": "string",
    },
    "enum": {
      "enum": [
        "Salmon",
        "Tuna",
        "Trout",
      ],
      "type": "string",
    },
    "from": {
      "format": "date-time",
      "nullable": true,
      "type": "string",
    },
    "gt": {
      "exclusiveMinimum": true,
      "minimum": 5,
      "type": "number",
    },
    "gte": {
      "minimum": 5,
      "type": "number",
    },
    "includes": {
      "type": "string",
    },
    "int": {
      "type": "integer",
    },
    "intersection": {
      "allOf": [
        {
          "properties": {
            "name": {
              "type": "string",
            },
          },
          "required": [
            "name",
          ],
          "type": "object",
        },
        {
          "properties": {
            "role": {
              "type": "string",
            },
          },
          "required": [
            "role",
          ],
          "type": "object",
        },
      ],
    },
    "ipv4": {
      "format": "ipv4",
      "type": "string",
    },
    "ipv6": {
      "format": "ipv6",
      "type": "string",
    },
    "length": {
      "maxLength": 5,
      "minLength": 5,
      "type": "string",
    },
    "lookup": {
      "additionalProperties": {
        "type": "boolean",
      },
      "type": "object",
    },
    "lt": {
      "exclusiveMaximum": true,
      "maximum": 5,
      "type": "number",
    },
    "lte": {
      "maximum": 5,
      "type": "number",
    },
    "max": {
      "maxLength": 3,
      "type": "string",
    },
    "min": {
      "minLength": 10,
      "type": "string",
    },
    "multiple": {
      "default": "test@test.cl",
      "example": "sample@email.com",
      "format": "email",
      "maxLength": 12,
      "type": "string",
    },
    "multipleOf": {
      "multipleOf": 5,
      "type": "number",
    },
    "nan": {
      "default": NaN,
      "nullable": true,
      "readOnly": true,
      "type": "number",
    },
    "negative": {
      "exclusiveMaximum": true,
      "maximum": 0,
      "type": "number",
    },
    "never": {
      "readOnly": true,
    },
    "nonnegative": {
      "minimum": 0,
      "type": "number",
    },
    "nonpositive": {
      "maximum": 0,
      "type": "number",
    },
    "null": {
      "default": null,
      "format": "null",
      "nullable": true,
      "type": "number",
    },
    "nullable": {
      "nullable": true,
      "type": "boolean",
    },
    "nullish": {
      "nullable": true,
      "type": "boolean",
    },
    "number": {
      "type": "number",
    },
    "numberOrBool": {
      "oneOf": [
        {
          "type": "number",
        },
        {
          "type": "boolean",
        },
      ],
    },
    "object": {
      "properties": {
        "id": {
          "exclusiveMinimum": true,
          "minimum": 0,
          "type": "number",
        },
        "name": {
          "type": "string",
        },
      },
      "required": [
        "id",
        "name",
      ],
      "type": "object",
    },
    "optional": {
      "type": "boolean",
    },
    "positive": {
      "exclusiveMinimum": true,
      "minimum": 0,
      "type": "number",
    },
    "record": {
      "additionalProperties": {
        "type": "number",
      },
      "type": "object",
    },
    "regex": {
      "pattern": "[a-c]",
      "type": "string",
    },
    "startsWith": {
      "type": "string",
    },
    "string": {
      "type": "string",
    },
    "stringOrNumber": {
      "oneOf": [
        {
          "type": "string",
        },
        {
          "type": "number",
        },
      ],
    },
    "tru": {
      "enum": [
        true,
      ],
      "type": "boolean",
    },
    "tuna": {
      "enum": [
        "tuna",
      ],
      "type": "string",
    },
    "tuple": {
      "items": {
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "boolean",
          },
        ],
      },
      "maxItems": 2,
      "minItems": 2,
      "type": "array",
    },
    "twelve": {
      "enum": [
        12,
      ],
      "type": "number",
    },
    "ulid": {
      "example": "01ARZ3NDEKTSV4RRFFQ69G5FAV",
      "format": "ulid",
      "type": "string",
    },
    "undefined": {
      "readOnly": true,
    },
    "unknown": {
      "nullable": true,
    },
    "until": {
      "format": "date-time",
      "nullable": true,
      "type": "string",
    },
    "url": {
      "example": "http://www.example.com/",
      "format": "uri",
      "type": "string",
    },
    "uuid": {
      "type": "string",
    },
    "void": {
      "readOnly": true,
    },
  },
  "required": [
    "string",
    "number",
    "boolean",
    "null",
    "coerceDate",
    "tuna",
    "twelve",
    "tru",
    "max",
    "min",
    "length",
    "email",
    "url",
    "emoji",
    "uuid",
    "cuid",
    "cuid2",
    "ulid",
    "regex",
    "includes",
    "startsWith",
    "endsWith",
    "ipv4",
    "ipv6",
    "gt",
    "gte",
    "lt",
    "lte",
    "int",
    "positive",
    "nonnegative",
    "negative",
    "nonpositive",
    "multipleOf",
    "nan",
    "datetime",
    "from",
    "until",
    "enum",
    "nullable",
    "object",
    "array",
    "arrayNonEmpty",
    "arrayOf5",
    "arrayFrom5",
    "tuple",
    "stringOrNumber",
    "numberOrBool",
    "discriminated",
    "record",
    "lookup",
    "intersection",
  ],
  "type": "object",
}
`;
